Fix manual renewal subscriptions that use SEPA when disabling the Legacy experience

a-danae opened the pull request:
Fixes #3133

## Changes proposed in this Pull Request:

- Introduce a migrator class extending [WCS_Background_Repairer](https://github.com/Automattic/woocommerce-subscriptions-core/blob/trunk/includes/abstracts/abstract-wcs-background-repairer.php)
- Select the subscriptions using legacy SEPA tokens (those with the `stripe_sepa` gateway ID) to be migrated
- For each subscription, create an updated SEPA gateway token if none exists. 
    This isn't needed to fix the problem, but doing it for consistency for the shoppers. Assuming it could be strange for shoppers to see an active subscription but no tokens.
- Update the subscription payment method type to use the updated SEPA `stripe_sepa_debit` gateway ID. 
    This is what fixes the actual problem

## Testing instructions

ℹ️ Since 8.2.0, legacy SEPA payment methods aren't attached to the customer for some reason. Check out 8.1.0 to add legacy SEPA payment methods if you want to test processing renewals. I'll create an issue to check this out.

### Setup

- Install and activate [WC Subscriptions](https://woocommerce.com/products/woocommerce-subscriptions/)
- Create a subscription product
- Create a customer user to be used as the shopper
- Set the store currency to Euro 
- Create a page using the classic checkout if you don't have one already
- Set up webhooks
- Under the payment methods tab, enable SEPA. At `wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=methods`

### Create subscriptions with the Legacy SEPA gateway

1. Log in as the shopper
1. Go to My Account > Payment methods > Add payment method, at `/my-account/add-payment-method/`
1. Add three SEPA payment methods. 
    [Find test IBAN numbers here](https://docs.stripe.com/sources/sepa-debit#testing). `AT611904300234573201`, `DE89370400440532013000`, `EE382200221020145685`
1. Add a card payment method 
1. As the shopper, purchase the subscription product four times - one with each payment method.
    You'll need to use the shortcode checkout. Legacy SEPA doesn't show up in Blocks
1. Confirm that
    - All subscriptions are active
    - Each of them display the selected payment method
    <img width="700" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/3aa7e2b2-1627-462f-81cc-b35d87eaaaf8">
1. Go to the Edit page for each subscription at `/wp-admin/admin.php?page=wc-orders--shop_subscription&action=edit&id=< sub ID >`
1. Under "Payment method", confirm SEPA Direct Debit is displayed
    <img width="700" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/6306ea89-d51c-40ef-8cc5-e6133c0d0264">

At this point:
- Going into the database > the wc_orders table > these SEPA subscriptions will have `stripe_sepa` under the `payment_method` column
- The Legacy SEPA payment gateway is enabled on the Stripe extension, which uses the `stripe_sepa` gateway ID 

### See the issue

1. Disable the Legacy experience
1. As a merchant, go to the Edit page for each subscription
1. Notice that under "Payment method", it now says "Manual renewal"

At this point:
- The Legacy SEPA payment gateway (`stripe_sepa`) is disabled
- The Updated SEPA payment gateway (`stripe_sepa_debit`) is enabled instead
- WooCommerce flags the payment gateway used for the subscription is disabled, thus Subscription changes the subscription to Manual renewal

### Confirm the right subscriptions are queued to be updated - actually testing the fix now

1. As a merchant, go to the Scheduled Action page
1. Confirm that there's a scheduled action called `wc_stripe_schedule_subscriptions_legacy_sepa_token_repairs`
    <img width="700" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/b308f7fe-2356-4213-8ab9-06a765c2b139"> 
1. Click on "Run now", or wait for the action to run
    This action will schedule a new action for each subscription that needs to be updated
1. If you let it execute on its own, confirm that another `wc_stripe_schedule_subscriptions_legacy_sepa_token_repairs` doesn't get scheduled for later
1. Search for actions called `wc_stripe_subscriptions_legacy_sepa_token_repair`
1. Confirm there are three actions scheduled. 
1. Confirm the ID of the subscription each one of them targets belongs to a subscription purchased with the Legacy SEPA gateway
1. Confirm that the subscription purchased with the card isn't listed
    <img width="700" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/932c291e-9c4a-48f4-855a-f3a29c796496">


### Confirm the subscriptions' payment method gets updated

1. As a merchant, go to the Scheduled Action page
1. Search for actions called `wc_stripe_subscriptions_legacy_sepa_token_repair`
1. Run the actions for each subscription. Or, preferably, wait for them to run
1. As a merchant, go to the Edit page for each subscription
1. Confirm that under "Payment method", it now says "SEPA Direct Debit"
1. On the WooCommerce logs, at `wp-admin/admin.php?page=wc-status&tab=logs`, confirm that there are entries for each subscription that was updated

What we're doing for each subscription:
- We retrieve all of the `stripe_sepa_debit` tokens for the user that owns the subscription. 
- When calling `WC_Payment_Tokens::get_customer_tokens()`, we create tokens using the Updated SEPA gateway (`stripe_sepa_debit`) from the Legacy tokens (`stripe_sepa`), [over here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8.3.0/includes/class-wc-stripe-payment-tokens.php#L316)
- We update the `payment_method` of the subscription, from `stripe_sepa` to `stripe_sepa_debit`, which matches the gateway ID of the Updated SEPA gateway
- WooCommerce no longer flags the payment gateway used for the subscription as disabled; thus Subscription is no longer flagged under "Manual renewal"


### Unhappy paths

- No action must be scheduled when WC_Subscriptions isn't active
- No action must be scheduled when the Legacy experience is enabled

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

a-danae committed:
Add base migrator class to update the subscriptions' payment method

a-danae committed:
Retrieve the batch of subscriptions to be updated

a-danae committed:
Add hook properties to the migrator class

a-danae committed:
Retrieve the token associated with the subscription during migration

a-danae committed:
Reorder migrator methods based on their visibility

a-danae committed:
Use the default token when either the legacy or updated tokens could be retrieved

a-danae committed:
Create a token with the updated SEPA gateway from the legacy token when migrating

a-danae committed:
Handle retrieving the updated token in a separate method

a-danae committed:
Merge branch 'develop' into fix/manual-renewal-subscriptions-with-sepa

a-danae committed:
Use the existing functionality behind get_customer_tokens to create updated SEPA tokens

a-danae committed:
Retrieve only the SEPA tokens for the customer

a-danae committed:
Remove inline comment

a-danae commented on includes/migrations/class-wc-stripe-subscriptions-legacy-sepa-tokens-migrator.php:
```diff
@@ -0,0 +1,200 @@
+	 * @param WC_Subscription  $subscription Subscription for which the payment method must be updated.
+	 * @param WC_Payment_Token $token        The token to be set as the payment method for the subscription.
+	 */
+	private function set_subscription_updated_payment_method( WC_Subscription $subscription, WC_Payment_Token $token ) {
```
Hey @mattallan, @james-allan ! Quick question here. Would this be the right way to update the payment method associated with the subscription?

This is what we do in [WC_Stripe_Subscriptions_Trait::maybe_update_source_on_subscription_order()](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L427), but wondering if we should use [WCS_Payment_Tokens::update_subscription_token()](https://github.com/Automattic/woocommerce-subscriptions-core/blob/trunk/includes/class-wcs-payment-tokens.php#L30) instead, or [WC_Order::add_payment_token()](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php#L842).

The renewals work fine when updating the payment method like this, but maybe I'm missing something.

mattallan replied:
Good question! What you have here is perfectly fine and IMO a bit better than just using `WCS_Payment_Tokens::update_subscription_token()` which doesn't call `set_payment_method()` to ensure the correct Stripe SEPA gateway ID is stored as the subscription.

Using the `WCS_Payment_Tokens::update_subscription_token()` will still work and has some added benefits like updating the latest renewal order with the updated token and also trigger the `'woocommerce_subscription_token_changed'` but I don't think these are necessary to the migrator, are they?

I did a quick search for 'woocommerce_subscription_token_changed' across Woo and Automattic orgs and can't see any plugins using it so I think it's good to go with what you have here.

Curious what @james-allan thinks!

james-allan replied:
I think `set_payment_method()` is good. It's the one we also use in things like the REST API.

a-danae committed:
Pass the logger instance as a parameter of the migrator class

a-danae committed:
Update the visibility of the get_items_to_repair method

a-danae committed:
Use a log file specific for upgrading the tokens

a-danae committed:
Paginate the results when retrieving subscriptions to update

a-danae committed:
Update scheduled and repair hooks from the updater class

a-danae committed:
Add method for conditionally scheduling repairing subscriptions

a-danae committed:
Add context in comments to the migrator class

a-danae committed:
Reorder the get_items_to_repair method within the class based on its visiblity

a-danae committed:
Update the conditionals for updating a subscription

With this change, the subscription meta for flagging that the migration took place won't skip the migration. Also, we're including the check for whether the Updated checkout experience is enabled into the method that would trigger an exception when the migration is skipped, so the reason gets logged.

a-danae committed:
Update the visibility of the maybe_update_subscription_legacy_payment_method method

a-danae committed:
Initialize the updating of subscriptions when updating the extension version

a-danae committed:
Merge branch 'develop' into fix/manual-renewal-subscriptions-with-sepa

a-danae committed:
Rename the class for updating the subscriptions' tokens

a-danae committed:
Add PHP unit test base file for the legacy SEPA token updater

a-danae committed:
Add php unit tests for retrieving the items to repair

a-danae committed:
Add php unit tests for exceptions when reparing an item

a-danae committed:
Add an entry to the log once the subscription gets updated

a-danae committed:
Add php unit tests for handling the token to be updated

a-danae committed:
Add tests for updating the legacy token with the same source ID of an updated one

a-danae committed:
Use the class property instead of a string multiple times

a-danae committed:
Add helper class to mock WC_Subscriptions_Cart in PHP unit tests

a-danae committed:
Fix unexpected payment type value when the order has a subscription

a-danae commented on tests/phpunit/test-class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -1537,6 +1537,9 @@ public function test_if_order_has_subscription_payment_method_will_be_saved() {
 		list( $amount, $description, $metadata ) = $this->get_order_details( $order );
 
+		// When the order contains a subscription, the payment type is expected to be "recurring".
+		$metadata['payment_type'] = 'recurring';
```
This test isn't related to the introduced changes. It started to fail in [f9cf21f](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3139/commits/f9cf21f574d5c6e88cf53b073e395ccf4023fcf4) due to the introduction of the [WC_Subscriptions test helper class](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3139/commits/f9cf21f574d5c6e88cf53b073e395ccf4023fcf4#diff-5c210fbd4f0b20f01c9cd83dd44302bc82a332b9e22f4cbe72e28f991192b883).

We check whether the payment is recurring [when setting up](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8.3.0/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php#L1694) this metadata, calling [is_payment_recurring()](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8.3.0/includes/compat/trait-wc-stripe-subscriptions-utilities.php#L52-L66). Before creating the test helper class, [is_subscriptions_enabled()](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8.3.0/includes/compat/trait-wc-stripe-subscriptions-utilities.php#L14-L23) likely returned `false` due to the `class_exists( 'WC_Subscriptions' )` check. This would make the previous assertion pass, even though it's not what would be expected.

a-danae committed:
Move the test helper class WCS_Background_Repairer to its own file

a-danae committed:
Always require the class for updating the sepa tokens

a-danae committed:
Require the WCS_Background_Repairer heleper class before the main extension in tests

a-danae committed:
Always initialize the subscriptions updater

Otherwise, the scheduled action doesn't have any attached callback and fails

a-danae committed:
Don't include the updater when Subscriptions isn't active

a-danae committed:
Remove checks for WC_Stripe_Subscriptions_Legacy_SEPA_Tokens_Update::init

This is now called every time outside WC_Stripe_Subscriptions_Legacy_SEPA_Tokens_Update's methods so scheduled actions have callbacks

a-danae committed:
Schedule the repairing action only once

a-danae committed:
Remove flag for updating sepa tokens on uninstall

a-danae committed:
Check for subscriptions using the Legacy SEPA token again on plugin update

a-danae committed:
Merge branch 'develop' into fix/manual-renewal-subscriptions-with-sepa

a-danae committed:
Create an updated token from the legacy one instead of looking for a default token

a-danae committed:
Only pass the source ID instead of the token for updating the subscription

a-danae committed:
Don't bail out when no token is found

We only need the source_id for the subscription to continue renewing, which it already has. We just need to update the payment method ID associated with it.

james-allan commented on includes/migrations/class-wc-stripe-subscriptions-legacy-sepa-tokens-update.php:
```diff
@@ -0,0 +1,271 @@
+	 * @return WC_Payment_Token|false
+	 */
+	private function get_customer_token_by_source_id( string $source_id, int $user_id, string $gateway_id ) {
+		$customer_tokens = WC_Payment_Tokens::get_customer_tokens( $user_id, $gateway_id );
```
I wonder if we should introduce a helper function for this kind of thing. I think I had to do something similar (query for a token by `pm_`), and WC doesn't support that. 

I other reason I ask is because there's the potential that if the customer has a lot of tokens (more than 10) this function may fail to locate the token because by default WC limits that query to 10. If that happens we then create a new token. Which will just further contribute to that problem.

a-danae replied:
> I wonder if we should introduce a helper function for this kind of thing. I think I had to do something similar (query for a token by pm_), and WC doesn't support that.

Sounds great, yeah. I was surprised we didn't have a function to find a token by its `token` value.

> I other reason I ask is because there's the potential that if the customer has a lot of tokens (more than 10) this function may fail to locate the token because by default WC limits that query to 10. If that happens we then create a new token. Which will just further contribute to that problem.

Right. I tried to reduce that risk by retrieving only the tokens for a specific gateway instead of retrieving all tokens. This way, we'd have to reach the query limit for tokens for this gateway for tokens to be left out. It's less likely that a user will have 10 SEPA tokens saved, but yes, still possible. 

As an alternative to reduce that possibility, we could also use the `woocommerce_get_customer_payment_tokens_limit` [filter](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-payment-tokens.php#L84C37-L84C82) to increase that limit. Although I agree a helper function would actually fix the potential problem.

Would this be a blocker for the PR?

james-allan commented on includes/migrations/class-wc-stripe-subscriptions-legacy-sepa-tokens-update.php:
```diff
@@ -0,0 +1,271 @@
+	 *
+	 * @param string  $source_id Source ID from which to create the new token.
+	 * @param integer $user_id
+	 * @return void
```
```suggestion
	 * @return WC_Payment_Token_SEPA|bool The new SEPA token, or false if no legacy token was found.
```

james-allan commented on includes/migrations/class-wc-stripe-subscriptions-legacy-sepa-tokens-update.php:
```diff
@@ -0,0 +1,271 @@
+		// If no updated SEPA token was found, create a new one based on the source ID.
+		if ( ! $replacement_token ) {
+			$replacement_token = $this->create_updated_sepa_token( $source_id, $user_id );
+		}
```
I might be confused here but this function reads as:

1. Find the token for the `$source_id`. 
2. If no token was found, create an updated one. 

But if I look at `create_updated_sepa_token()` it will first check if the token exists, and if it doesn't it will exit. 

Combined, doesn't that mean that if the legacy token doesn't exist, we call the create one, but it will skip because a legacy one doesn't exist?

james-allan replied:
I was confused. :)

`$replacement_token = $this->create_updated_sepa_token( ... );` here tries to find the token using the **new gateway ID** (`stripe_sepa_debit`), if it doesn't find a saved token with the new gateway ID and a matching `pm_` it will create a new one. 

`create_updated_sepa_token()` will then attempt to find an existing token using the old ID (`stripe_sepa`)

ie if there's no token for `stripe_sepa_debit` create one but only if there's one with the `stripe_sepa`.

a-danae replied:
That, yes 😄 Thanks for going through it again. Glad to update this method if it's not as clear as it should!

james-allan reviewed (commented):
Nice work here @a-danae! The attention to detail and quality of this PR is so nice. I've only just reviewed the code so far. I tried to start testing however I'm getting errors trying to use the sources APIs.

eg When I try to add a saved SEPA token I get this: 

<img width="536" alt="Screenshot 2024-06-11 at 4 32 17 PM" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/d2fcd874-c415-4b7f-8eb0-9b833c2eba06">

I'm not sure if there's a trick to get around this. None of the tips [here](https://support.stripe.com/questions/i-received-a-sources-api-error-how-should-i-handle-it) worked and none of my existing Stripe test accounts worked either. I may need to borrow some credentials. I've left some preliminary comments.

I also had some general questions about this migration: 

1. What happens if a merchant goes back to the legacy experience after disabling it? Do we have plans for that or do we expect that to be a irreversible migration? 
2. I don't think this is possible because of how the repairer schedules its jobs, however, if a subscription is renewed before the migrator has had a chance to migrate it. It would still be treated as manual I suspect. 

The reason I ask is because I wonder if we need to consider a Just-in-time kind of approach where if the data is accessed before we've had a chance to migrate it, we fix it then and there. I'm not sure if a neat solution exists for that. I suspect we'd need to do something similar to the workaround mentioned here: https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3133#issuecomment-2107643526

a-danae commented:
Big thanks for the review and the comments here @james-allan ! 🙌 

> When I try to add a saved SEPA token I get this:

Ah yeah, I've gotten this with new Stripe accounts. I've invited you to the Stripe account I use to test this. Looks like we need to use accounts that had processed payments with the Sources API [before](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows-%E2%80%90-Instructions#alternative-payment-methods-apms).

> 1. What happens if a merchant goes back to the legacy experience after disabling it? Do we have plans for that or do we expect that to be a irreversible migration?

For now, yes. After chatting with Diego, we agreed not to migrate back to the Legacy SEPA gateway for now because the Sources API will be deprecated. We'll want to fix whatever is making the merchant go back to Legacy, but we can implement the migration back if we get reports of merchants needing to migrate back. Glad to take a different approach for this, though.

> 2. I don't think this is possible because of how the repairer schedules its jobs, however, if a subscription is renewed before the migrator has had a chance to migrate it. It would still be treated as manual I suspect.

Good point... I suspect so, yep.

> The reason I ask is because I wonder if we need to consider a Just-in-time kind of approach where if the data is accessed before we've had a chance to migrate it, we fix it then and there. I'm not sure if a neat solution exists for that. I suspect we'd need to do something similar to the workaround mentioned here: https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3133#issuecomment-2107643526

Makes sense. I hadn't thought of it, other than that filter. Do you mean implementing both approaches?

james-allan commented:
> I've invited you to the Stripe account I use to test this.

That worked. Thanks for that! 

---

Verify the issue: 

| Before Disabling Legacy | After Disabling Legacy |
|--------|--------|
| <img width="858" alt="Screenshot 2024-06-12 at 2 42 29 PM" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/460a3b39-c33d-4ac2-b95d-5947d22c6b50"><p align="center"><sup>All SEPA/Credit card payment method</sup></p> | <img width="851" alt="Screenshot 2024-06-12 at 2 45 12 PM" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/cd84d3e9-b8e8-4272-864c-dcb1e6eb62ba"><p align="center"><sup>All SEPA subscriptions turned into **manual renewal**</sup></p> | 

I've confirmed that a scheduled action was created for all 4 impacted SEPA subscriptions. The one with a credit card is excluded. 

<img width="1282" alt="Screenshot 2024-06-12 at 3 05 52 PM" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/d3640314-a6ed-4686-a75e-0a4696f53c4e">

> [!note]
In [this comment](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3139#pullrequestreview-2109411779) I mentioned that because of the way the repairer schedules these actions, it shouldn't be possible to run into an issue where a subscription tries to renew before we've had a chance to upgrade. I had forgotten that it schedules these jobs to run in 1 hour, rather than immediately. It's now pretty possible (although unlikely) to run a renewal within that 1 hour period. 

Once those actions were completed, the subscriptions were correctly updated to `stripe_sepa_debit`.

| Actions | Subscriptions |
|--------|--------|
| <img width="808" alt="Screenshot 2024-06-12 at 3 59 48 PM" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/4b3d395d-5fcd-46f4-b745-02f6f11437c1"> | <img width="929" alt="Screenshot 2024-06-12 at 4 00 55 PM" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/daae665f-0aed-4398-97fe-0536a02b9484"> | 

🕺 🕺 🕺 

----
A couple of other things I verified: 

- if you disable automatic renewals from the Subscriptions settings (accept manual renewals and disable automatic renewals), the subscription is still migrated and the subscription remains manual. 👍 
- the old payment method is stored in `_migrated_sepa_payment_method` meta. 

<p align="center">
<img width="488" alt="Screenshot 2024-06-12 at 4 16 27 PM" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/11c4e30f-d43a-42d4-a72e-c4e77abedad8">
</p>

- The logs are logging the migration information.
- ❌ I tested a couple of the conditions. eg:
   - _I purchased a SEPA subscription_
   - _disabled legacy experience_
   - _waited for it to schedule jobs_ 
   - _As a customer I changed the subscription's payment method (Add payment from the View Subscription page on the My Account) to a card PM_.
   - _I ran the repair job and it migrated the subscription back to SEPA._ 
   - You could add a condition to `get_subscription_to_migrate()` that throws an exception if the subscription is no longer set to SEPA. ie `---- Skipping migration of subscription #%d. The subscription is no longer set to SEPA.
- Disabling WooCommerce Subscriptions and the there's no fatal errors and the repairer isn't loaded. 👍 
- If you reenable the legacy experience before the repair jobs start running, then the jobs are skipped. With helpful logs for troubleshooting.

<p align="center">
<img width="923" alt="Screenshot 2024-06-12 at 4 38 01 PM" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/f3a310b4-1463-4aa2-8ae5-f24d1b819ca7">
</p>

> Makes sense. I hadn't thought of it, other than that filter. Do you mean implementing both approaches?

Yeah so I initially had concerns with the approach of using a background repair script because if there is a simpler approach, it's almost always better to go with that. After testing this PR, I don't really have those concerns anymore given this repair script is doing more than just changing the payment method. I've also come around to the fact that this repair script approach is necessary because search-ability is affected. eg if you use the payment method filter on the list table to locate subscriptions with SEPA, it would return no results until this migration occurs. 

Another thought I had to avoid these subscriptions being treated as manual was to hook into the subscription renewal action and migrate the subscription before the subscription attempts to renew. For example, if you hook onto `woocommerce_scheduled_subscription_payment` at 0, check if the subscription is `stripe_sepa` and if so migrate it in time for the renewal. 

I think both approaches would supplement each other. We do want to update the data in the database without having to wait until the renewal occurs - which is the job of this repairer. To avoid missed renewals we could hook onto  `woocommerce_scheduled_subscription_payment` and migrate any just-in-time. 

📓 I still need to test the recreated/updating token behavior of this PR.
I also want to consider if there's any improvements we can make to the loading of this class. ie when we know we're done, stop loading it.

a-danae committed:
Update includes/migrations/class-wc-stripe-subscriptions-legacy-sepa-tokens-update.php

Co-authored-by: James Allan <james.allan@automattic.com>

a-danae commented:
Thanks for another round of reviews and insights @james-allan ! 

> I ran the repair job and it migrated the subscription back to SEPA.

Great catch! ~Updating this~ Updated in [0f03a27](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3139/commits/0f03a277e96e8693655513ee45a34fc69d2756b4).

> I also want to consider if there's any improvements we can make to the loading of this class. ie when we know we're done, stop loading it.

Makes sense. Checking. Would this be a blocker?

> Another thought I had to avoid these subscriptions being treated as manual was to hook into the subscription renewal action and migrate the subscription before the subscription attempts to renew. For example, if you hook onto woocommerce_scheduled_subscription_payment at 0, check if the subscription is stripe_sepa and if so migrate it in time for the renewal.

Hadn't thought of that, sounds great. Would it be okay to implement this approach in a separate PR? Thinking it'd be an scenario that could be tested separately, and thinking of not making this PR bigger and not having more testing steps for it.

a-danae committed:
Merge branch 'develop' into fix/manual-renewal-subscriptions-with-sepa

a-danae committed:
Skip repairing a subscription when the associated payment method isn't legacy SEPA

a-danae commented:
@james-allan , ~just a heads up that I re-requested your review, but I'd like to adjust if following your comments before reviewing it again. I marked this as a draft in the meantime.~ Ok, marking this as ready to review. Planning to address the comments in separate PRs.

a-danae committed:
Rename the repairer class from the subscriptions Legacy SEPA tokens

a-danae committed:
Separate the background repairer logic from the actual updating of the subscriptions

a-danae commented on includes/compat/class-wc-stripe-subscriptions-legacy-sepa-token-update.php:
```diff
@@ -0,0 +1,184 @@
+ *
+ * This class extends the WCS_Background_Repairer for scheduling and running the individual migration actions.
+ */
+class WC_Stripe_Subscriptions_Legacy_SEPA_Token_Update {
```
I'm separating the logic for updating the token from the background repairer class to be reused in the `woocommerce_scheduled_subscription_payment` action. Planning to do that in a separate PR.

a-danae committed:
Update methods docblocks

james-allan commented:
> > I also want to consider if there's any improvements we can make to the loading of this class. ie when we know we're done, stop loading it.
>
> Makes sense. Checking. Would this be a blocker?

Nah, not a blocker.

> Hadn't thought of that, sounds great. Would it be okay to implement this approach in a separate PR? Thinking it'd be an scenario that could be tested separately, and thinking of not making this PR bigger and not having more testing steps for it.

Yeah that sounds like a good idea, this PR is tough to test so it would be good to move those additional things off to separate PRs.

james-allan commented on includes/migrations/class-wc-stripe-subscriptions-legacy-sepa-tokens-update.php:
```diff
@@ -187,6 +188,12 @@ private function get_subscription_to_migrate( $subscription_id ) {
 			throw new \Exception( sprintf( '---- Skipping migration of subscription #%d. Subscription not found.', $subscription_id ) );
 		}
 
+		$subscription_payment_method = $subscription->get_payment_method();
```
This can be deleted. This variable isn't used. :)

james-allan commented on includes/compat/class-wc-stripe-subscriptions-legacy-sepa-token-update.php:
```diff
@@ -0,0 +1,190 @@
+			throw new \Exception( sprintf( '---- Skipping migration of subscription #%d. Subscription not found.', $subscription_id ) );
+		}
+
+		$subscription_payment_method = $subscription->get_payment_method();
```
```suggestion
```

This can be removed as this variable isn't used. :)

james-allan reviewed (approved):
The changes tested well again today. I left a minor comment but I'm approving this PR as it's not blocking and it would be good to progress this further.

james-allan commented:
@a-danae I also wanted to flag that after the payment method is migrated, I tried to renew the subscription, however I got this error from Stripe.

<img width="526" alt="Screenshot 2024-06-21 at 1 11 53 PM" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/839e087e-82aa-46fd-a21f-75632f6e0133">

What I think is causing that is we're using the payment intents API to try collect a payment with a src payment method.

But I'm not 100% sure. Just wanted to flag so you were aware.

james-allan commented:
> But I'm not 100% sure. Just wanted to flag so you were aware.

It could also have something to do with the fact that my `src_` payment methods have a status of `consumed` and so they aren't chargeable anymore. I'm not sure what's behind that.

a-danae committed:
Update includes/compat/class-wc-stripe-subscriptions-legacy-sepa-token-update.php

Co-authored-by: James Allan <james.allan@automattic.com>

a-danae committed:
Remove extra empty line

a-danae commented:
Summarizing here the pending items mentioned in the review that will be addressed separately:

- [ ] Create a helper function to retrieve a token for a customer by the token ID. [Ref](
https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3139#discussion_r1634247721)
- [ ] Address the scenario when a shopper has more tokens than the posts_per_page limit. Related to the previous point. [Ref](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3139#discussion_r1643452815)
- [x] Prevent loading the repairing class after all affected subscriptions were repaired. [Ref](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3139#issuecomment-2174502742)
   GH Issue https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3238
- [x] Investigate the `invalid_request_error` error when processing renewals of repaired subscriptions. [Ref](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3139#issuecomment-2181930597) https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3242
- [x] Repair the subscription during renewal, in case the repairing hasn’t happened by then. [Ref](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3139#issuecomment-2174502742) https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3243

a-danae committed:
Merge branch 'develop' into fix/manual-renewal-subscriptions-with-sepa

a-danae committed:
Add changelog entries

james-allan commented:
> Summarizing here the pending items mentioned in the review that will be addressed separately:

Are there any you want some help with @a-danae?

agenceKanvas commented:
I'm a little scared to try it actually. Is there a way to launch the migrator on 4 ou 5 subscriptions and see if it's working on production before launching it globally ?

a-danae commented:
> Are there any you want some help with @a-danae?

That'd be great! I'm creating issues for these points now so we can distribute them. Big thanks @james-allan !

> Is there a way to launch the migrator on 4 ou 5 subscriptions and see if it's working on production before launching it globally ?

@agenceKanvas sure, it makes sense to test with a few of them first. We don't have a way to control them right now, but adding a filter when retrieving the items to repair would do the trick; it'd allow merchants to control this if needed and it would be an easy addition.

@james-allan happy to go differently if you disagree! Thinking of adding a filter to the return value of [get_items_to_update()](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3139/files#diff-5dd518b55598de74e10c77faf65baffca430aa74882b57cb4864b8b3ddcfe5b2R98)

james-allan commented:
> @james-allan happy to go differently if you disagree! Thinking of adding a filter to the return value of [get_items_to_update()](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3139/files#diff-5dd518b55598de74e10c77faf65baffca430aa74882b57cb4864b8b3ddcfe5b2R98)

We'd need to be careful not to set the option that then prevented it from running for the remaining subscriptions or I guess we could just include a note that after running you're test you need to delete the option to trigger it to run for the remaining ones.

agenceKanvas commented:
@a-danae , I've tried the testing scenario on my staging site and the wc_stripe_schedule_subscriptions_legacy_sepa_token_repairs did not appear (v8.5.0-test). 
Is there a way to call it manually ?

james-allan commented:
> Is there a way to call it manually ?

@agenceKanvas if you have access to run code directly, something like this will kick it off.

```php
$logger  = wc_get_logger();
$updater = new WC_Stripe_Subscriptions_Repairer_Legacy_SEPA_Tokens( $logger );
$updater->schedule_repair();
```

> [!important]
> Once you run it, you will want to remove this code because you don't want multiple instances of the `WC_Stripe_Subscriptions_Repairer_Legacy_SEPA_Tokens` class running on your site.

a-danae commented:
> (v8.5.0-test).

Hey @agenceKanvas, just a heads up that [Stripe 8.5.1 is out](https://github.com/woocommerce/woocommerce-gateway-stripe/releases/tag/8.5.1) and contains this fix, so you can use that instead of the test release in GH.

agenceKanvas commented:
Hi @a-danae ! 
I've tried to run the commands. This is the result : 
- Commands runned with no error displayed
- Event listed in the scheduled one, run manually and finished
- The 3 subscriptions that I created following the procedure were still in manual renewal state
I'm going to import the db export that I made before the command, update the plugin and check again. 
If that's not working, what data would you need to see what could have happen ? 
It's a staging server so if you need access I can provide one. 
Thanks for your help !

agenceKanvas commented:
So, I reinstalled the backup of the database, before the plugin update, I installed the 8.5.1, and the event was not triggered again. 
I tried to run it manually with success : 
![Capture d’écran 2024-07-15 à 11 09 34](https://github.com/user-attachments/assets/8e58a6ab-1309-43e5-8d34-dd04973dcb37)

But the subs are sill in manual renewal :
![Capture d’écran 2024-07-15 à 11 12 11](https://github.com/user-attachments/assets/900e0c88-c19e-47f7-888b-e7246571d97c)

** Update **
I spoke too fast. When I go in the DB and query the post meta related to payment of the sub, I have what I'm supposed to have : 
![Capture d’écran 2024-07-15 à 15 41 40](https://github.com/user-attachments/assets/458974cd-c1c4-47cb-8330-f646e5f5e5b4)
The only problem, I think, is that the next payment date outdated. 
Because this is a staging website, the payments are locked in manual renewal. 
When it goes live, will the date be updated automatically ? 
Thanks a lot !

a-danae commented:
Hi there @agenceKanvas ! Glad to hear the meta was updated as expected

Not sure if you got this sorted, but just in case...

> When it goes live, will the date be updated automatically ?

The date wouldn't be updated as we only change the payment-related meta for the migration. Have you encountered any issues with payment renewal dates having passed? We also hook the migration on renewals, so if it's on Manual renewal and the renewal date comes, the payment will go through and the payment method will be updated for upcoming renewals.
